Support for MultiGriddedData

dulte opened the pull request:
## Change Summary
Adds possibility to read multiple model files, e.g. uEMEP gridded city simulation. This is done through the new MultiGriddedData, which holds a list of GriddedData files. As of now, only functions that are needed for colocation using colocate_gridded_ungridded are supported.

While trying to have as small a footprint as possible, some changes to the ungridded container was needed, to make sure the regridding works correctly for both gridded and ungridded data.

Assumes model file have the same projection (or lack of projection), vars, units, ts_type and ndim. The data must also be disjoint (no overlapping).

Is tested with uEMEP and EMEP data

- Full run started:
  - Output will (eventually) be available here: https://aeroval-test.met.no/thlun/pages/evaluation/?project=uemep&experiment=uemep-2025-08-29
  - Output dir: `/lustre/storeB/users/thlun8736/python/aeroval/data/uemep/uemep-2025-08-29`

## Checklist

* [x] Start with a draft-PR
* [x] The PR title is a good summary of the changes
* [x] PR is set to AeroTools and a tentative milestone
* [x] Documentation reflects the changes where applicable
* [x] Tests for the changes exist where applicable
* [x] Tests pass locally
* [x] Tests pass on CI
* [x] At least 1 reviewer is selected
* [x] Make PR ready to review

dulte committed:
First try at multigrid

dulte committed:
Fixes x/yrange for selection stations

dulte committed:
adds city regions

dulte committed:
Adds latlon filter with more than one range to ungridded

dulte committed:
Starts tests for multigrid

dulte committed:
Adds more tests for multigriddeddata, as well as other changes made to pyaerocom

dulte committed:
Fixes multigriddeddata fixture

dulte committed:
removes some fixtures that caused problems

dulte committed:
Reverts change that broke some tests

dulte committed:
adds newest testdata

dulte committed:
Changes used testdata

dulte committed:
Adds extra test for multigridded

dulte committed:
Adds docstrings, and removes dead code

dulte committed:
Fixes test

dulte committed:
Merge branch 'main-dev' into multigrid

dulte committed:
Changes ALLOWED_VERT_COORD_TYPES to ALLOWED_COORD_TYPES in filter_by_latlon

dulte committed:
Changes from MultiGriddedData to GriddedDataContainer

dulte committed:
Removes some unused code, and linting

dulte committed:
Fixes some of the broken test. Tests having to do with gridded obs are still not working

dulte commented:
Tests fail for gridded_gridded (as expected), and for aeroval/utils. The functions in aeroval/utils are never used, so will probably remove this code and its tests :wastebasket:

dulte committed:
Fixed some tests, broken others...

dulte committed:
Fixes some tests. Add a decorator to griddeddata_container for checking if only one child is present

dulte committed:
removes dead code

thorbjoernl committed:
Try to fix

thorbjoernl committed:
Merge branch 'main-dev' into multigrid-merge

thorbjoernl commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,552 @@
+            self.add_griddeddata(data)
+
+    def _create_latlon_info(self, data: GriddedData) -> dict:
+        return {
```
1. This should deal with standard name not existing as some uEMEP files don't have it (?).

2. Not sure if this is an issue on my end, but when I ran uEMEP against EBAS in EMEP config, the dimensions are lat / lon, not latitude / longitude. Maybe something needs to be more flexible or there is an issue somewhere further up.

thorbjoernl commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,552 @@
+        # self._lat_res = data.lat_res
+        # self._lon_res = data.lon_res
+
+        self._lat_points = data.latitude.points
+        self._lon_points = data.longitude.points
```
Needs to also account for lat / lon names; see other comment.

thorbjoernl commented on pyaerocom/griddeddata.py:
```diff
@@ -993,6 +993,69 @@ def transpose(self, new_order):
+
+        """
+
+        latitude = self.latitude.points
+        longitude = self.longitude.points
+        lat_range = (np.min(latitude), np.max(latitude))
```
This will also need to take into account lat / lon possibility

thorbjoernl committed:
Small fixes

thorbjoernl committed:
Allow overwriting emep var_map instead of just updating.

thorbjoernl committed:
.

thorbjoernl committed:
Don't use lat/lon - use projection

thorbjoernl committed:
fix GriddedDataContainer + no proj_info case

thorbjoernl committed:
wip

thorbjoernl committed:
Fix cams283 test failure

thorbjoernl committed:
.

thorbjoernl committed:
Remove code comments

thorbjoernl committed:
Remove more code comments

thorbjoernl committed:
ditto

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -771,25 +767,25 @@ def latlon_proj(lat, lon):
+            longitude = data.longitude.points
+            xrange = [np.min(longitude), np.max(longitude)]
+            yrange = [np.min(latitude), np.max(latitude)]
+        elif isinstance(data, GriddedDataContainer):
```
Would be nice to have a `get_lonlat_ranges()` in a GriddedData/GriddedDataContainer and use it just here.

(Maybe we have that already for data with projections? `data.get_xyranges()` ?)

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,499 @@
+    pass
+
+
+class GriddedDataContainer:
```
Would be nice to have a common interface for GriddedDataContainer/GriddedData so it is clear which functions have to exist in both.

dulte replied:
I'll make that now that I have an overview over all the functions that needs to exist in both

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,499 @@
+class GriddedDataContainer:
+    def __init__(self, data_id: str, data: GriddedData | None = None):
+        """
+        Class for holding and working with multiple GriddedData objects.
```
What is the difference between `MultiGriddedData` and `GriddedDataContainer`? Are both in use?

dulte replied:
No. GriddedDataContainer is meant to take over from MultiGriddedData. So I will remove all traces of the latter

thorbjoernl committed:
Remove MultiGriddedData

heikoklein commented on tests/aeroval/test__processing_base.py:
```diff
@@ -2,7 +2,7 @@
 import pytest
 
-from pyaerocom import Colocator, GriddedData, UngriddedData
+from pyaerocom import Colocator, UngriddedData, GriddedDataContainer
```
Should only import UngriddedDataContainer interface and test against it, not one of the implementations, e.g. UngriddedData

heikoklein reviewed (approved):
Looks good. Some ideas of improving the interface, but that can also be done at a later stage.

heikoklein reviewed (changes requested):
The evaluation of https://aeroval-test.met.no/thlun/pages/evaluation/?project=uemep&parameter=concNno2&observation=EBAS&station=Cabauw+Wielsekade&experiment=uemep-2025-08-29# show again the wrong values of Cabauw Wielsekade, (1.09 ug N/m3 instead of 4.38 ug N/m3 as in the experiment 2025-08-28-2)

heikoklein commented on pyaerocom/griddeddata.py:
```diff
@@ -531,7 +531,7 @@ def lon_res(self):
         vals = np.diff(self.longitude.points)
         val = vals.mean()
-        if vals.std() / val > 0.0001:
+        if vals.std() / val > 0.1:  # NOTE: Increased from 0.0001
```
why is this needed for multigridded data with projection. Checking the resolution of the 2d longitude grid does not make any sense. For lat/lon data without projection, the noise-to-signal ratio should not be changed.

dulte replied:
This is a problem/fix for the uEMEP 250m. When using 250m gridding this check is not passed. I think this is due to the more irregular coords to make it fit with the gridding. It might be errors in the precision of the lat/lon or something. I need to figure out how to handle the fine resolution, and then I will change this check back

heikoklein replied:
Is this still needed? You promised to change it back...

thorbjoernl commented:
# Status

- Current config I'm working with: `/lustre/storeB/project/fou/kl/emep/Reports/2025/SourceReceptor/Evaluation/cfg_emep-status_reporting2025_uemep.py`
	- I've left some explanatory comments. See comments starting with `NOTE: `
- Most recent evaluations: 
	- https://aeroval-test.met.no/thlun/pages/evaluation/?project=uemep&experiment=uemep-2025-08-29
	- https://aeroval-test.met.no/thlun/pages/evaluation/?project=uemep&experiment=uemep-2025-08-29-2
	- The only difference is that the *second* evaluation is done on only the Netherlands (`uemep/157`, `uemep/158`) tiles.
		- **Problem:** Wrong model values for stations and regions. Since the including of extra tiles changes the station values, I suspect that data that shouldn't be included is included. May be some operation that doesn't use proj_info correctly.
			- The issue is already present in the collocated data afaict so it happens during colocation.

dulte committed:
Adds coord sorting to to_time_series in griddeddata_container

dulte committed:
Adds different behaviour for griddeddata container with one child and one with more children

dulte commented:
to_time_series now makes sure that the resulting array is sorted in the same way as the incoming coords, and thus the stations in the station list for obs and mod in `colocate_gridded_ungridded` should be the same

dulte committed:
Removes interpolation on each coord, and adds sorting of stationdata on coords instead

dulte commented:
@heikoklein, I've removed the interpolation of single coords we talked about, and instead implemented the simplest possible sorting of the list of stationdatas, to see if that worked

This will fail for e.g. LCS, where multiple stations can have the same coords :( So I will look into your idea (colocating each tile separately) when I'm back  

Here is the result: https://aeroval-test.met.no/danielh/pages/evaluation/?project=uemep&statistic=nmb&experiment=uemep-full-2&region=ALL#

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -369,69 +369,31 @@ def to_time_series(
+            lat = float(stationdata["latitude"])
+            lon = float(stationdata["longitude"])
 
+            dists2 = [(point[0] - lat) ** 2 + (point[1] - lon) ** 2 for point in points]
```
You don't know here if you have x/y coordinates or lat/lon coordinates. A eukledian metric as used her is only valid for x/y coordinates, while you have to use in general the haversine formula to get it working for lat/lon.

The general problem is not that to_time_series destroys the order, but that you only get the positions of tiles, and the tiles don't have a order. You know already from the bounding-box test earlier which stations belong to which tile. You should try not loose that information (in colocation_utils.py I think).

heikoklein replied:
Please remove `sort_stationdatas_by_coords()`. It's no longer needed as far as I can see.

dulte replied:
removed

dulte committed:
Adds sorting of station based on which tile they belong, instead of sorting model stations to colocation

dulte commented:
Due to the failings of the above sorting, I've switched to sorting obs stations based on which tile they belong. 

I tried to colocate each tile separately, but that increased the runtime to much (due to having to filter the obs once per tile). The new sorting is a compromise between the first sorting method and colocation each tile separately. 

https://aeroval-test.met.no/danielh/pages/evaluation/?project=uemep&statistic=data_mean&experiment=uemep-full-3&region=ALL#

dulte committed:
Adds new tests, with new testdata-minimal containing lcs

dulte committed:
Corrects default testdata

dulte committed:
Merge branch 'main-dev' into multigrid

dulte committed:
Adds pyaro-readers to pyproject

dulte committed:
Removes pyaro-readers from dep

dulte committed:
Adds extra test for split model file and griddeddata_container

dulte committed:
adds pyaro-readers to test deps

dulte committed:
Updates version of pyaro-readers used

dulte committed:
Changes how the sorted is done for tiles

dulte commented:
The last commit lowers the runtime for LCS and uemep from ~24 hours to about 1.5 hours

thorbjoernl commented:
[Here](https://aeroval-test.met.no/thlun/pages/evaluation/?project=multigrid-test-final) is a new status run and full uemep evaluation with the most recent push.

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,553 @@
+                latitude = data.latitude.points
+                longitude = data.longitude.points
+                xrange = [np.min(longitude), np.max(longitude)]
+                yrange = [np.min(latitude), np.max(latitude)]
```
in hires/uemep, latitude and longitude can be huge. And latitude/longitude is in lat-lon projection guaranteed to be monotonous. So use data.latitude[0]/[-1] and get min/max of those.

heikoklein replied:
Why do we have this min/max implementation so often? Should only exist for lat/lon and x/y, i.e. twice.

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,553 @@
+                xrange = [np.min(longitude), np.max(longitude)]
+                yrange = [np.min(latitude), np.max(latitude)]
+            else:
+                xrange, yrange = self._get_xyrange_child(data)
```
confusing name with `_child` since you are in the child-loop already. `_get_xyrange` or `_get_xyrange_data`

dulte replied:
Changed name to just `_get_xyrange`

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,553 @@
+            yranges.append(yrange)
+        return xranges, yranges
+
+    def _get_xyrange_child(self, child: GriddedData) -> tuple[tuple[float, float]] | tuple[None]:
```
This method is designed to be independent of self, and just work on "child". Make it a staticmethod. (and change child to data)

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,553 @@
+        for coord in child.cube.dim_coords:
+            if coord.var_name == child.proj_info.x_axis:
+                vals = coord.points
+                xrange = (np.min(vals), np.max(vals))
```
Don't use min/max on all values, which are guaranteed to be monotonic.

Do we have access to axes bounds? Currently we might loose half grid-cells to the right and left.

dulte replied:
I've changed it to `xrange = (vals[0], vals[-1])`. The axes are not guaranteed to have bounds, sadly

heikoklein replied:
Can you document in the docstrings what you mean by xrange then, i.e. that it is the ordered frist and last value of the axes, and not the beginning and end of the data-range? Just to be clear. (Or you could check for bounds, use them if they exist, and if not, assume value is center and extrapolate the range half a cell...)

dulte replied:
This code was originally in `colocation_utils.py`, and has been there the last year. (I just moved it to be a method of griddeddata_container) So if we are going to change it's behavior (i.e. use bounds or extrapolate) then we might want to have an own PR for that, not let it be part of this big PR

heikoklein replied:
A docstring saying what we actually do would be nice.

heikoklein commented on pyaerocom/griddeddata_container.py:
```diff
@@ -0,0 +1,553 @@
+                xrange = (np.min(vals), np.max(vals))
+            if coord.var_name == child.proj_info.y_axis:
+                vals = coord.points
+                yrange = (np.min(vals), np.max(vals))
```
Same as above

dulte replied:
see above

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -971,14 +962,60 @@ def latlon_proj(lat, lon):
+            [*proj(station.latitude, station.longitude), station.latitude, station.longitude]
+            for station in unsorted_obs_stat_data
+        ],
+    )
```
obs_stat_pos has 4 axes in float and those are addressed by 0-3. Would be much nicer to have x,y,lat,lon indexes:
```python
np.array(
    [(*proj(...), lat, lon) for ...],
    np.dtype([('x', 'f8'), ('y', 'f8'), ('lat', 'f8'),('lon', 'f8')])
)
```
and access then as `obs_stat_pos["x"]` rather than `obs_stat_pos[:, 0]`

dulte replied:
Implemented this

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -971,14 +962,60 @@ def latlon_proj(lat, lon):
+
+        found_id = np.where(
+            np.logical_and(
+                np.logical_and(obs_stat_pos[:, 0] >= xrange[0], obs_stat_pos[:, 0] < xrange[1]),
```
why do we exclude xrange[1]? Ranges are currently not bounds but last grid-point included. Change to bounds (see construction of range) or include.

heikoklein replied:
Please add a comment that xrange is inclusive.

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -971,14 +962,60 @@ def latlon_proj(lat, lon):
+                np.logical_and(obs_stat_pos[:, 0] >= xrange[0], obs_stat_pos[:, 0] < xrange[1]),
+                np.logical_and(obs_stat_pos[:, 1] >= yrange[0], obs_stat_pos[:, 1] < yrange[1]),
+            )
+        )[0]
```
Is this the same as
```python
xrange[0] <= obs_stat_pos['x'] & obs_stat_pos['x']  <= xrange[1] &
yrange[0] <= obs_stat_pos['y'] & obs_stat_pos['y'] <= yrange[1]
```
`np.logical_and` is the same as &, and `np.where` does not do anything on a logical array?
```

dulte replied:
The where is included so that the it's the indices that are returned instead of the bool array. This is done because I couldn't find a working way of having `unsorted_obs_stat_data` (see below)

heikoklein replied:
I think `&` is shorter/nicer than `np.logical_and`, and writing bound-checks in natural order, e.g. `(xrange[0] <= val) & (val <= xrange[1])` is much easier to read than your code. (in non-numpy code this would be even nicer, e.g. `xrange[0] <= val <= xrange[1]` but my example is very close to that)

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -971,14 +962,60 @@ def latlon_proj(lat, lon):
+            )
+        )[0]
+
+        found_stations = [unsorted_obs_stat_data[i] for i in found_id]
```
`unsorted_obs_stat_data[found_id]` ? Or am I missing something? Please comment

dulte replied:
It is because it is a list, and thus I can't pass in a list/array of indices. I tried passing in `unsorted_obs_stat_data` as an array, but it it did not preserve the elements correctly as StationData, so I opted for a list

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -971,14 +962,60 @@ def latlon_proj(lat, lon):
+
+        found_stations = [unsorted_obs_stat_data[i] for i in found_id]
+
+        ungridded_lats = list(obs_stat_pos[found_id, 2])
```
why convert to python-list?
`obs_stat_pos['lat'][found_id]`

heikoklein commented on pyaerocom/colocation/colocation_utils.py:
```diff
@@ -971,14 +962,60 @@ def latlon_proj(lat, lon):
+        ungridded_lats = list(obs_stat_pos[found_id, 2])
+        ungridded_lons = list(obs_stat_pos[found_id, 3])
+
+        if len(ungridded_lats) == 0:
```
abort earlier, as soon as you have done the bounds-check

heikoklein replied:
not addressed in your review. Could be moved right after construction of found_id

heikoklein commented on pyaerocom/griddeddata.py:
```diff
@@ -993,6 +993,69 @@ def transpose(self, new_order):
+        for coord in self.cube.dim_coords:
+            if coord.var_name == self.proj_info.x_axis:
+                vals = coord.points
+                xrange = (np.min(vals), np.max(vals))
```
don't use min/max but [0] and [-1] (monotonic), only check if [0] is larger or smaller than [-1]

heikoklein commented on pyaerocom/griddeddata.py:
```diff
@@ -993,6 +993,69 @@ def transpose(self, new_order):
+        latitude = self.latitude.points
+        longitude = self.longitude.points
+        lat_range = (np.min(latitude), np.max(latitude))
+        lon_range = (np.min(longitude), np.max(longitude))
```
if lat/lon projection/1D, use [0]/[-1] rather than whole array.

heikoklein commented on pyaerocom/projection_information.py:
```diff
@@ -110,3 +110,15 @@ def from_xarray(ds: xarray.Dataset, var: str):
+            return False
+        if value._units != self._units:
+            return False
+        return True
```
Slightly dangerous because projections might be equal, even if the attribute strings are now, but ok.
Why not use python boilerplate code her:
```python
    def __eq__(self, other):
        if not isinstance(other, type(self)):
            return False
        return self.__dict__ == other.__dict__
```

heikoklein commented on tests/test_griddeddata_container.py:
```diff
@@ -0,0 +1,175 @@
+    for model in ["uEMEP", "EMEP"]:
+        data = cities_data[model]
+        for city in data:
+            assert isinstance(city, GriddedData)
```
Can't you test against `GriddedDataContainer` here?

dulte replied:
No, I'm checking if all children are GriddedData

heikoklein commented on tests/test_griddeddata_container.py:
```diff
@@ -0,0 +1,175 @@
+from copy import deepcopy
+
+from pyaerocom.griddeddata_container import GriddedDataContainer, GriddedDataContainerException
+from pyaerocom import GriddedData
```
`GriddedData` really needed?

dulte replied:
See above

heikoklein reviewed (changes requested):
Great PR, fast implementation and new functionality.

Some nitbits

dulte committed:
Addresses first part of review

dulte committed:
Continues addressing the review

dulte commented:
I think I've addressed all the points now. For some reason the 3.10 tests fail. I think it has to do with the version of typer. I had the bug with some other project

dulte committed:
Addesses more of the review

heikoklein commented:
I just merged a fix from @charlienegri to get the tests working again with newer xarray and older typer. Please try merging main-dev to this branch.

dulte committed:
Merge branch 'main-dev' into multigrid

dulte committed:
Removes to_time_series from griddeddata_container

dulte committed:
Skips lat_res test due to poor test data

dulte commented:
lat_res seems to work fine with uEMEP reporting data. So the test will be skipped for now
